### PR TITLE
Add validations and tests for core classes

### PR DESCRIPTION
**Pull request by neonowy**
cc @gotar

### CODE REVIEW

File: lib/basket.rb [hunk 1]
```diff
@@ -10,24 +12,29 @@ def initialize(warehouse)
...
+
     if warehouse.in_stock?(product)
       item = find_or_create_item(product)
 
       item.increase_quantity
 
       warehouse.decrease_product_status(product)
     else
-      puts "Sorry, product is not available at the moment :("
+      raise ProductNotFound
```
Comment (gotar) [inline lib/basket.rb:24]: 👍

File: lib/basket_item.rb [hunk 2]
```diff
@@ -11,6 +11,8 @@ def increase_quantity
   end
 
   def decrease_quantity
+    raise StandardError unless @quantity > 0
```
Comment (gotar) [inline lib/basket_item.rb]: will be better to add specific error to describe it, like:

`QuantityLevelError = Class.new(StandardError)`

File: lib/invoice.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    invoice += get_header
+
+    @basket.items.each_with_index do |item, index|
+      invoice += get_basket_item(item, index)
+    end
+
+    invoice += get_summary
+
+    invoice
```
Comment (gotar) [inline lib/invoice.rb]: read about `tap` method in ruby, will fit here perfectly

Comment (neonowy) [inline lib/invoice.rb]: Do you mean something like this?

``` ruby
def get_invoice
  invoice = ""

  invoice.tap do |i|
    i << get_header

    @basket.items.each_with_index do |item, index|
      i << get_basket_item(item, index)
    end

    i << get_summary
  end
end
```

Comment (gotar) [inline lib/invoice.rb]: more like:

``` ruby
def get_invoice
  "".tap do |invoice|
    ...
  end
end
```

Comment (gotar) [inline lib/invoice.rb]: name can be better then `get_invoice`, it suggest you download it from somewhere, or fetch from db or disk, but you generate it - `generate_invoice`, `prepare_invoice`, `render_invoice` maybe will be better - please rethink it

**Comment by neonowy** [comment]
I resolved all the issues, I think.

cc @gotar

File: Gemfile [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+source "https://rubygems.org"
+
+gem "rspec"
```
Comment (gotar) [inline Gemfile]: you should add rspec inside `group :test`

File: lib/invoice.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+  def generate_header
+    header = ""
+
+    header += "Rachunek dla:\n"
+    header += "#{@buyer}\n"
+    header += "------------------\n\n"
+
+    header
```
Comment (gotar) [inline lib/invoice.rb]: you can use `tap` here too

File: lib/invoice.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    sum = '%.2f' % @basket.sum
+    sum_with_vat = '%.2f' % @basket.sum_with_vat
+
+    summary += "\n"
+    summary += "------------------\n"
+    summary += "  Suma: #{sum} PLN\n"
+    summary += " z VAT: #{sum_with_vat} PLN\n"
+
+    summary
```
Comment (gotar) [inline lib/invoice.rb]: you can use `tap` here too

File: lib/user.rb [hunk 7]
```diff
@@ -17,9 +17,9 @@ def remove_from_basket(product)
     basket.remove(product)
   end
 
-  def print_invoice
+  def generate_invoice
     invoice = Invoice.new(buyer: name, basket: basket)
 
-    invoice.print
+    invoice.generate_invoice
```
Comment (gotar) [inline lib/user.rb]: No need to set var for that you can just:

``` ruby
Invoice.new(buyer: name, basket: basket).generate_invoice
```

File: lib/basket.rb [hunk 8]
```diff
@@ -63,4 +70,4 @@ def find_item(product)
   def find_item_index(product)
     @items.index { |i| i.product == product }
```
Comment (mmarcinkowski) [inline lib/basket.rb:71]: u can try to use `find_index` method instead -> http://ruby-doc.org/core-2.3.1/Array.html#method-i-find_index

File: spec/unit/basket_item_spec.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
+require_relative "../../lib/basket_item"
+require_relative "../../lib/product"
+
+RSpec.describe BasketItem do
+  subject(:basket_item) { BasketItem.new(product) }
+  let(:product) { Product.new(name: "Ball", price: 10) }
```
Comment (gotar) [inline spec/unit/basket_item_spec.rb]: you should move price to `let`, and then in test do not write: `60`, just `6*price`

File: lib/basket.rb (hunk 8 above)
Comment (neonowy) [inline lib/basket.rb:71]: Hmm.. How it's different from `index` method?

**Pull request merged**